File: Crafting Interpreters/jlox/src/java/main/com/craftinginterpreters/tool/GenerateAst.java
package com.craftinginterpreters.tool;

import java.io.IOException;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.List;

public class GenerateAst {
  public static void main(String[] args) throws IOException {
    if (args.length != 1) {
      System.err.println("Usage: generate_ast <output director>");
      System.exit(1);
    }

    defineAst(
        args[0],
        "Expr",
        Arrays.asList(
            "Binary   : Expr left, Token operator, Expr right",
            "Grouping : Expr expression",
            "Literal  : Object value",
            "Unary    : Token operator, Expr right"));
  }

  private static void defineAst(String outputDir, String baseName, List<String> types)
      throws IOException {
    String path = outputDir + "/" + baseName + ".java";
    PrintWriter writer = new PrintWriter(path, "UTF-8");

    writer.println("package com.craftinginterpreters.lox;");
    writer.println("");
    writer.println("import java.util.List;");
    writer.println("");
    writer.printf("abstract class %s {\n", baseName);

    defineVisitor(writer, baseName, types);

    // AST classes.
    for (String type : types) {
      String className = type.split(":")[0].trim();
      String fields = type.split(":")[1].trim();
      defineType(writer, baseName, className, fields);
    }

    writer.println("\n  abstract <R> R accept(Visitor<R> visitor);");
    writer.println("}");
    writer.close();
  }

  private static void defineType(
      PrintWriter writer, String baseName, String className, String fields) {
    // Start of class definition
    writer.println("");
    writer.printf("  static class %s extends %s {\n", className, baseName);

    // Constructor
    String[] fieldList = fields.split(",");
    writer.printf("    %s (%s) {\n", className, fields);
    for (String field : fieldList) {
      String name = field.trim().split(" ")[1].trim();
      writer.printf("      this.%s = %s;\n", name, name);
    }
    writer.printf("    }\n\n");

    // Visitor
    writer.println("\n    <R> R accept(Visitor<R> visitor) {");
    writer.printf("      return visitor.visit%s%s(this);\n", className, baseName);
    writer.println("    }");

    // Fields
    for (String field : fieldList) {
      writer.printf("    final %s;\n", field);
    }

    // End of class definition
    writer.printf("  }");
  }

  private static void defineVisitor(PrintWriter writer, String baseName, List<String> types) {
    writer.println("    interface Visitor<R> {");

    for (String type : types) {
      String typeName = type.split(":")[0].trim();
      writer.printf(
          "      R visit%s%s (%s %s);\n", typeName, baseName, typeName, baseName.toLowerCase());
    }

    writer.println("    }");
  }
}

File: Crafting Interpreters/src/main/scala/tool/GenerateAst.scala
package com.craftinginterpreters.lox.tool

import java.io.PrintWriter

object GenerateAst extends App {
  if (args.length != 1) {
    System.err.println("Usage: scala src/main/scala/tool/GenerateAst.scala <directory>")
    System.exit(1)
  }

  defineAst(args(0), "Stmt", Array(
    "Expression - expression: Expr",
    "Print      - expression: Expr",
    "Var        - name: Token, initializer: Expr"
  ))

  defineAst(args(0), "Expr", Array(
    "Binary   - left: Expr, operator: Token, right: Expr",
    "Grouping - expression: Expr",
    "Literal  - value: Any",
    "Unary    - operator: Token, right: Expr",
    "Variable - name: Token"
  ))

  def defineAst(outputDir: String, baseName: String, types: Array[String]) = {
    val path = s"${outputDir}/${baseName}.scala"
    val writer = new PrintWriter(path, "UTF-8")

    println(s"Generating $path.")

    defineHeader(writer)
    writer.println("")

    defineImports(writer)
    writer.println("")

    defineClass(writer, baseName)
    writer.println("")

    defineObject(writer, baseName, types)
    writer.close()
  }

  def defineHeader(writer: PrintWriter) = {
    writer.println("/* AUTO GENERATED - DO NOT EDIT */")
    writer.println("package com.craftinginterpreters.lox")
  }

  def defineImports(writer: PrintWriter) = {
    writer.println("import collection.mutable.MutableList")
  }

  def defineObject(writer: PrintWriter, baseName: String, types: Array[String]) = {
    writer.println(s"object ${baseName} {")
    defineVisitor(writer, baseName, types)

    types.foreach { ttype =>
      val className = ttype.split("-")(0).trim()
      val fields = ttype.split("-")(1).trim()
      writer.println("")
      defineType(writer, baseName, className, fields)
    }

    writer.println("}")
  }

  def defineClass(writer: PrintWriter, baseName: String) = {
    writer.println(s"abstract class ${baseName} {")
    writer.println(s"  def accept[T](visitor: ${baseName}.Visitor[T]): T")
    writer.println("}")
  }

  def defineType(writer: PrintWriter, baseName: String, className: String, fields: String) = {
    val justArgs = fields
      .split(",")
      .map { field => field.split(":")(0).trim() }
      .mkString(", ")

    val accessArgs = fields
      .split(",")
      .map { field => field.trim }
      .map { field => s"val $field" }
      .mkString(", ")

    // Constructor
    writer.println(s"  class $className($accessArgs) extends $baseName {")

    // Visitor
    writer.println(s"    def accept[T](visitor: Visitor[T]): T = {")
    writer.println(s"      visitor.visit${className}${baseName}(new $className($justArgs))")
    writer.println("    }")
    writer.println("  }")
  }

  def defineVisitor(writer: PrintWriter, baseName: String, types: Array[String]) = {
    writer.println("  trait Visitor[T] {")

    for (ttype <- types) {
      val typeName = ttype.split("-")(0).trim()
      writer.println(s"    def visit${typeName}${baseName} (${baseName.toLowerCase}: ${typeName}): T")
    }

    writer.println("  }")
  }
}
